network.c

```diff
@@ -70,85 +74,92 @@ static int _set_netmask(int skfd, char *intf, char *newmask)
+			}
+		}
+	}
 
-	for (ifa = ifaddr, n = 0; ifa != NULL; ifa = ifa->ifa_next, n++) {
-		if (ifa->ifa_addr == NULL)
-			continue;
+	return iface;
+}
+
+void pv_network_update_meta(struct pantavisor *pv)
+{
+	struct ifaddrs *ifaddr = _get_sort_ifaces();
 
-		family = ifa->ifa_addr->sa_family;
-		if (family == AF_PACKET)
+	struct pv_json_ser js;
+	pv_json_ser_init(&js, 512);
+	pv_json_ser_object(&js);
+
+	char last[IFACE_KEY_SIZE] = { 0 };
+	char cur[IFACE_KEY_SIZE] = { 0 };
+	char host[NI_MAXHOST] = { 0 };
+
+	for (struct ifaddrs *ifa = ifaddr; ifa; ifa = ifa->ifa_next) {
+		if (!ifa->ifa_addr)
 			continue;
 
-		getnameinfo(ifa->ifa_addr,
-			    (family == AF_INET) ? sizeof(struct sockaddr_in) :
-							sizeof(struct sockaddr_in6),
-			    host, NI_MAXHOST, NULL, 0, NI_NUMERICHOST);
-
-		SNPRINTF_WTRUNC(iff, sizeof(iff), "%s.%s", ifa->ifa_name,
-				family == AF_INET ? "ipv4" : "ipv6");
-		if (!strcmp(ifn, iff)) {
-			ilen += strlen(host) + 4;
-			ifaddrs = realloc(ifaddrs, ilen);
-			t = strdup(ifaddrs);
-			SNPRINTF_WTRUNC(ifaddrs, ilen, "%s,\"%s\"", t, host);
-			free(t);
+		char *family = NULL;
+		size_t size = 0;
+		if (ifa->ifa_addr->sa_family == AF_PACKET) {
+			continue;
+		} else if (ifa->ifa_addr->sa_family == AF_INET) {
+			family = "ipv4";
+			size = sizeof(struct sockaddr_in);
 		} else {
-			SNPRINTF_WTRUNC(ifn, sizeof(ifn), "%s.%s",
-					ifa->ifa_name,
-					family == AF_INET ? "ipv4" : "ipv6");
-			ilen = 0;
-			ilen += strlen(host) + 4;
-			ifaddrs = realloc(ifaddrs, ilen);
-			SNPRINTF_WTRUNC(ifaddrs, ilen, "\"%s\"", host);
-		}
-		if (ifa->ifa_next != NULL) {
-			SNPRINTF_WTRUNC(iff, sizeof(iff), "%s.%s",
-					ifa->ifa_next->ifa_name,
-					ifa->ifa_next->ifa_addr->sa_family ==
-							AF_INET ?
-						      "ip4" :
-						      "ipv6");
-
-			if (!strncmp(ifn, iff, sizeof(ifn)))
-				continue;
+			family = "ipv6";
+			size = sizeof(struct sockaddr_in6);
 		}
 
-		SNPRINTF_WTRUNC(ifn, sizeof(ifn), "%s.%s", ifa->ifa_name,
-				family == AF_INET ? "ipv4" : "ipv6");
-		size = sizeof(IFACE_FMT) + strlen(ifn) + strlen(ifaddrs);
-		buf = calloc(size, sizeof(char));
-		len += snprintf(buf, size, IFACE_FMT, ifn, ifaddrs);
-		len++;
-		ifaces = realloc(ifaces, len);
-		strcat(ifaces, buf);
-		free(buf);
-		if (ifa->ifa_next != NULL)
-			strcat(ifaces, ",");
-	}
-	free(ifaddrs);
+		getnameinfo(ifa->ifa_addr, size, host, NI_MAXHOST, NULL, 0,
+			    NI_NUMERICHOST);
+
+		snprintf(cur, IFACE_KEY_SIZE, "%s.%s", ifa->ifa_name, family);
+		if (strncmp(cur, last, IFACE_KEY_SIZE)) {
+			if (strlen(last) > 0)
+				pv_json_ser_array_pop(&js);
 
-	ifaces = realloc(ifaces, len + 1);
-	strcat(ifaces, "}");
+			strncpy(last, cur, IFACE_KEY_SIZE);
+			pv_json_ser_key(&js, last);
+			pv_json_ser_array(&js);
+		}
+		pv_json_ser_string(&js, host);
+	}
+	pv_json_ser_array_pop(&js);
+	pv_json_ser_object_pop(&js);
 
-	pv_metadata_add_devmeta(DEVMETA_KEY_INTERFACES, ifaces);
+	pv_metadata_add_devmeta(DEVMETA_KEY_INTERFACES, pv_json_ser_str(&js));
```

Comment:
you have to free the output of pv_json_ser_str (maybe add that in a comment in utils/json.h so we don't forget next time)